sklearn/neighbors/loop.py

```diff
@@ -0,0 +1,447 @@
+        X : array-like, shape (n_samples, n_features), default=None
+            The query sample or samples to compute the Local Outlier
+            Probability w.r.t. to the training samples.
+
+        Returns
+        -------
+        is_inlier : array, shape (n_samples,)
+            Returns -1 for anomalies/outliers and 1 for inliers.
+        """
+
+        return self.fit(X)._predict()
+
+    def fit(self, X, y=None):
+        """Fit the model using X as training data.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix, BallTree, KDTree}
+            Training data. If array or matrix, shape [n_samples, n_features],
+            or [n_samples, n_samples] if metric='precomputed'.
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+        if not (0. < self.norm_factor <= 1.):
+            raise ValueError("norm_factor must be in (0, 1.0]")
+
+        super(LocalOutlierProbability, self).fit(X)
+
+        LocalOutlierMixin.fit(self, X, y, mode='loop')
+
+        return self
+
+    def _predict(self, X=None):
+        """Predict the labels (1 inlier, -1 outlier) of X according to LoOP.
+
+        If X is None, returns the same as fit_predict(X_train).
+        This method allows to generalize prediction to new observations (not
+        in the training set). As LoOP originally does not deal with new data,
+        this method is kept private.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features), default=None
+            The query sample or samples to compute the Local Outlier
+            Probability w.r.t. to the training samples. If None, makes
+            prediction on the training data without considering them as
+            their own neighbors.
+
+        Returns
+        -------
+        is_inlier : array, shape (n_samples,)
+            Returns -1 for anomalies/outliers and +1 for inliers.
+        """
+        check_is_fitted(self,
+                        ["threshold_", "negative_local_outlier_probability_",
+                         "n_neighbors_", "_distances_fit_X_"])
+
+        is_inlier = LocalOutlierMixin._score_samples(self, X, mode='loop')
+
+        return is_inlier
+
+    @staticmethod
+    def _ssd(distances_X):
+        """
+        Calculates the sum of square distance of X. As this method is used in
+        the calculation of the local probability of each sample, this method
+        is kept private.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            The pairwise distances between each sample in the input data.
+
+        Returns
+        -------
+        ssd_array : array, shape (n_samples,)
+            Returns the sum of square distance of X as an array of
+            repeated values.
+        """
+        ssd_array = np.sum(np.power(distances_X, 2), axis=1)
+
+        return ssd_array
+
+    def _standard_distances(self, ssd_X):
```

Comment:
Why is this not static, but `_ssd` is, and wouldn't it be better to combine both into one, as they are only ever called as _standard_distances(_ssd)?